### PR TITLE
Fix issue #1210

### PR DESCRIPTION
**Pull request by mikeljordan**
## Proposed changes
 This PR addresses issue #1210 by refactoring the params dictionary into two more specific dictionaries: model_params and solver_params. This change improves code clarity and modularity by distinctly separating model configuration from solver settings.

**Key Changes:**
**Refactoring**: The params dictionary has been split into model_params and solver_params to provide a clearer distinction between model-specific parameters and solver-related configurations.
**Test and Tutorial Updates**: The changes have been applied where most needed, particularly in tests and tutorials, ensuring that the new parameter structure is consistently used.
**Default Parameters**: In the run_time_dependent_model() function, a default value has been introduced for solver_params, which is initialized in a safe manner to prevent duplications and ensure the function behaves as expected without requiring unnecessary parameter repetition.

## Types of changes
What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ x] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [ ] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [ x] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: src/porepy/models/run_models.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def run_stationary_model(model, params: dict) -> None:
     model.after_simulation()
 
 
-def run_time_dependent_model(model, params: dict) -> None:
+def run_time_dependent_model(model, params: Optional[dict] = None) -> None:
```
Comment (jwboth) [inline src/porepy/models/run_models.py:59]: Please also update the documentation. I would remove the question.

File: tests/functional/test_mandel.py [hunk 2]
```diff
@@ -150,12 +150,12 @@ def test_scaled_vs_unscaled_systems():
...
     }
     time_manager_unscaled = pp.TimeManager([0, 10], 10, True)
-    params_unscaled = {
+    model_params_unscaled = {
         "material_constants": material_constants_unscaled,
         "time_manager": time_manager_unscaled,
     }
-    unscaled = MandelSetup(params=params_unscaled)
-    pp.run_time_dependent_model(model=unscaled, params=params_unscaled)
+    unscaled_model = MandelSetup(params=model_params_unscaled)
```
Comment (jwboth) [inline tests/functional/test_mandel.py:157]: I would suggest the same logic in the naming for both the params dictionary as well as the model. So if using `model_params_unscaled` I would also go for `model_unscaled`. Maybe this is just my taste?

File: tests/functional/test_mandel.py [hunk 3]
```diff
@@ -165,22 +165,22 @@ def test_scaled_vs_unscaled_systems():
...
     units = pp.Units(**scaling)
-    params_scaled = {
+    model_params_scaled = {
         "material_constants": material_constants_scaled,
         "time_manager": time_manager_scaled,
         "units": units,
     }
-    scaled = MandelSetup(params=params_scaled)
-    pp.run_time_dependent_model(model=scaled, params=params_scaled)
+    scaled_model = MandelSetup(params=model_params_scaled)
```
Comment (jwboth) [inline tests/functional/test_mandel.py:173]: See comment above on naming.

File: tests/functional/test_terzaghi.py [hunk 4]
```diff
@@ -51,28 +51,28 @@ def test_biot_equal_to_incompressible_poromechanics():
...
             "fluid": pp.FluidConstants(terzaghi_fluid_constants),
         },
         "num_cells": 10,
     }
-    setup_poromech = TerzaghiSetupPoromechanics(params_poromech)
-    pp.run_time_dependent_model(setup_poromech, params_poromech)
+    setup_poromech = TerzaghiSetupPoromechanics(model_params_poromech)
+    pp.run_time_dependent_model(model=setup_poromech)
     p_poromechanics = setup_poromech.results[0].approx_pressure
     U_poromechanics = setup_poromech.results[0].approx_consolidation_degree
```
Comment (jwboth) [inline tests/functional/test_terzaghi.py:64]: We could clean up here, and use small `u_poromechanics`

File: tests/functional/test_terzaghi.py [hunk 5]
```diff
@@ -51,28 +51,28 @@ def test_biot_equal_to_incompressible_poromechanics():
...
             "fluid": pp.FluidConstants(terzaghi_fluid_constants),
         },
         "num_cells": 10,
     }
-    setup_biot = TerzaghiSetup(params_biot)
-    pp.run_time_dependent_model(setup_biot, params_biot)
+    setup_biot = TerzaghiSetup(model_params_biot)
+    pp.run_time_dependent_model(model=setup_biot)
     p_biot = setup_biot.results[0].approx_pressure
     U_biot = setup_biot.results[0].approx_consolidation_degree
```
Comment (jwboth) [inline tests/functional/test_terzaghi.py:77]: Same here -  small 'u'.

File: tests/models/test_boundary_condition.py [hunk 6]
```diff
@@ -249,16 +249,17 @@ class MomentumBalanceRobNeu(BCRobDir, MomentumBalanceNeu): ...
...
     }
 
     def _run_model(balance_class, alpha):
-        params["alpha"] = alpha
-        instance = balance_class(params)
-        pp.run_time_dependent_model(instance, params)
+        model_params["alpha"] = alpha
+        solver_params = {}
+        instance = balance_class(model_params)
+        pp.run_time_dependent_model(instance, params=solver_params)
```
Comment (jwboth) [inline tests/models/test_boundary_condition.py:262]: If I understood correctly, `solver_params` is obsolete here and you can call `run_time_dependent_model` without `params`?

File: tests/models/test_momentum_balance.py [hunk 7]
```diff
@@ -123,22 +123,22 @@ def test_unit_conversion(units, uy_north):
...
-    params = {
+    model_params = {
         "times_to_export": [],  # Suppress output for tests
         "fracture_indices": [0, 1],
         "cartesian": True,
         "uy_north": uy_north,
         "material_constants": {"solid": solid},
     }
-    reference_params = copy.deepcopy(params)
+    model_reference_params = copy.deepcopy(model_params)
```
Comment (jwboth) [inline tests/models/test_momentum_balance.py:133]: Why not `reference_model_params`? Wouldn't this better follow the style of calling the `params`, `model_params`?

File: tutorials/exporting_models.ipynb [hunk 8]
```diff
@@ -24,12 +24,14 @@
...
+    "model_params = {\n",
+    "    \"folder_name\": \"model_exporting\"\n",
+    "}\n",
+    "solver_params = {}\n",
     "\n",
     "# The compound class Poromechanics inherits from DataSavingMixin.\n",
-    "model_0 = Poromechanics(params)\n",
-    "pp.run_time_dependent_model(model_0, params)"
+    "model_0 = Poromechanics(model_params)\n",
+    "pp.run_time_dependent_model(model_0, solver_params)"
```
Comment (jwboth) [inline tutorials/exporting_models.ipynb:34]: Did you include `solver_params` on purpose?

File: tutorials/equations.ipynb [hunk 9]
```diff
@@ -40,14 +40,6 @@
    "execution_count": 1,
    "metadata": {},
    "outputs": [
-    {
```
Comment (jwboth) [inline tutorials/equations.ipynb:43]: The changes in this jupyter notebook are not really needed. Can you revert them?

File: tutorials/diagnostics.ipynb [hunk 10]
```diff
@@ -631,7 +572,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.11.4"
+   "version": "3.12.4"
```
Comment (jwboth) [inline tutorials/diagnostics.ipynb:575]: The changes in this notebook are not needed. Can you revert them?

File: tutorials/boundary_conditions.ipynb [hunk 11]
```diff
@@ -141,9 +141,10 @@
     "        pressure_vals[domain_sides.east] = self.fluid.convert_units(5, \"Pa\")\n",
     "        return pressure_vals\n",
     "\n",
-    "\n",
-    "single_phase_flow = SinglePhaseFlowExample1(params={})\n",
-    "pp.run_time_dependent_model(single_phase_flow, params={})\n",
+    "model_params = {}\n",
+    "solver_params = {}\n",
+    "single_phase_flow = SinglePhaseFlowExample1(params=model_params)\n",
+    "pp.run_time_dependent_model(single_phase_flow, params=solver_params)\n",
```
Comment (jwboth) [inline tutorials/boundary_conditions.ipynb:147]: Why using `solver_params` here?

File: tutorials/boundary_conditions.ipynb [hunk 12]
```diff
@@ -96,12 +96,12 @@
...
...wxfPrpp0KzZs2ENWvW6L0v8vPzdXXeeOMNIS0tTcjKyhLS09OF4OBgoV27dsLt27frZQzGjGP16tXCvn37hCtXrgjnz58XZs2aJVhZWQmHDx/W1bH016LC+PHjBT8/vyqPKcZrMWPGDMHOzk5IS0vT+x15+PChro4UPjOkrkkkJ4IgCB9++KHQsWNHwdraWhgwYIBw4sQJ3XODBw8WIiIi9Orv2rVLeOqppwRra2uhV69ewv79+/WeLysrE+bPny84OzsLCoVCeO6554TLly9bzBg8PDwEAJW2hQsXCoIgCA8fPhSGDh0qODo6Cs2bNxc8PDyE6dOn19sfLmPHMXv2bF1dZ2dnITQ0VDh79qze8Sz9tRAEQbh06ZIAQDh06FClY4n1WlRcjvrXrSL2iIgIYfDgwZXa+Pj4CNbW1kLnzp2FxMTESset6d9G7DEMHjy4xvqCUH55tKurq2BtbS20b99eGDNmjHD16tV6G4Mx41i+fLnQpUsXwcbGRmjTpo0QGBgofPPNN5WOa8mvhSCUX05ra2srrFu3rspjivFaVDUGAHq/61L5zJAymSAIQr1NyxARERHVUaM/54SIiIikhckJERERWRQmJ0RERGRRmJwQERGRRWFyQkRERBaFyQkRERFZFCYnREREZFGYnBAREZFFYXJCREREFoXJCREREVkUJidEVKM7d+7AxcUFS5cu1ZUdP34c1tbWlb6JlYjIHPjdOkRUqwMHDiA8PBzHjx+Hp6cnfHx8MHLkSKxatUrs0IioEWJyQkQGmTlzJg4fPoz+/fvj/Pnz+P7776FQKMQOi4gaISYnRGSQR48eoXfv3sjJycGZM2fQp08fsUMiokaK55wQkUGuXbuG3377DWVlZbh+/brY4RBRI8aZEyKqVUlJCQYMGAAfHx94enoiLi4O58+fh5OTk9ihEVEjxOSEiGo1Z84cfP755/jhhx/QqlUrDB48GHZ2dvjqq6/EDo2IGiEu6xBRjdLS0hAXF4ctW7ZAqVTCysoKW7ZswXfffYf4+HixwyOiRogzJ0RERGRROHNCREREFoXJCREREVkUJidERERkUZicEBERkUVhckJEREQWhckJERERWRQmJ0RERGRRmJwQERGRRWFyQkRERBaFyQkRERFZFCYnREREZFGYnBAREZFF+f+ByGZIPA3gMQAAAABJRU5ErkJggg==",
```
Comment (jwboth) [inline tutorials/boundary_conditions.ipynb:104]: Would it be possible to revert adding all the image data?

File: tutorials/boundary_conditions.ipynb [hunk 13]
```diff
@@ -270,10 +271,14 @@
...
+    "model_params = {\n",
+    "    \"material_constants\": {\"fluid\": fluid_constants}\n",
+    "}\n",
+    "solver_params = {}\n",
     "single_phase_flow = SinglePhaseFlowExample2(\n",
-    "    params={\"material_constants\": {\"fluid\": fluid_constants}}\n",
+    "    params=model_params\n",
     ")\n",
-    "pp.run_time_dependent_model(single_phase_flow, params={})\n",
+    "pp.run_time_dependent_model(single_phase_flow, params=solver_params)\n",
```
Comment (jwboth) [inline tutorials/boundary_conditions.ipynb:281]: Remove `solver_params`?

File: tutorials/grid_topology.ipynb [hunk 14]
```diff
@@ -409,7 +409,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.11.3"
+   "version": "3.12.4"
```
Comment (jwboth) [inline tutorials/grid_topology.ipynb:412]: The updates in this notebook seem irrelevant? Can you revert the changes?

File: tutorials/grids.ipynb [hunk 15]
```diff
@@ -32,7 +32,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
```
Comment (jwboth) [inline tutorials/grids.ipynb:35]: The updates in this notebook seem irrelevant. Can you revert the changes?

File: tutorials/boundary_conditions.ipynb [hunk 16]
```diff
@@ -385,11 +390,14 @@
...
+    "# Include the time_manager to the model params dictionary\n",
+    "model_params = {\n",
+    "    \"time_manager\": time_manager\n",
+    "}\n",
+    "solver_params = {}\n",
     "\n",
-    "model = MomentumBalanceTimeDependentBC(params)\n",
-    "pp.run_time_dependent_model(model=model, params=params)"
+    "model = MomentumBalanceTimeDependentBC(model_params)\n",
+    "pp.run_time_dependent_model(model=model, params=solver_params)"
```
Comment (jwboth) [inline tutorials/boundary_conditions.ipynb:400]: Remove `solver_params` from the arguments here?

File: tutorials/mandels_problem.ipynb [hunk 17]
```diff
@@ -168,76 +168,20 @@
...
-     "name": "stdout",
-     "output_type": "stream",
-     "text": [
-      "\n"
+     "ename": "NameError",
+     "evalue": "name 'solver_params' is not defined",
+     "output_type": "error",
+     "traceback": [
+      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
+      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
```
Comment (jwboth) [inline tutorials/mandels_problem.ipynb:176]: Could you revert the changes here (Error does not look nice :) ).

**Review by jwboth — COMMENTED** [review]
Thanks a lot. This looks pretty good already. 

Two comments:
- I see you did not go for changing the naming of the signature and call `params` also here `solver_params`. This is OK for me.
- I do not know what has been the practice with jupyter notebooks @keileg @IvarStefansson: Are we strict with massive changes as image data etc, when technically it is not needed? The same with execution counts, Python versions etc. 
- I was not able to see the diff for benchmark_simulation.ipynb and single_phase_flow.ipynb. I assume this is exactly because of updated image data. I would appreciate if you could revert the changes in this two, and only update the few lines with updated interface, e.g., using "git add -p" or some git GUI.

Please check the remaining (very mild) comments, answer as you wish and potentially update code where suited, and we will resolve the discussion subsequently. As you did a good job, I am confident the rest will go very smooth. Thanks again.

**Comment by keileg** [comment]
> * I see you did not go for changing the naming of the signature and call `params` also here `solver_params`. This is OK for me.

I agree, the context should make this clear.

> * I do not know what has been the practice with jupyter notebooks @keileg @IvarStefansson: Are we strict with massive changes as image data etc, when technically it is not needed? The same with execution counts, Python versions etc.

We have been very pragmatic on this one, partly because we rarely use the history for notebooks, partly because reading the text version of notebooks is a pain. In my opinion, we should focus on the rendered part, and let the metadata be what it is. Do you agree @IvarStefansson?

File: tutorials/boundary_conditions.ipynb [hunk 18]
```diff
@@ -14,7 +14,7 @@
   },
```
Comment (IvarStefansson) [inline tutorials/boundary_conditions.ipynb:14]: <pre class="ql-syntax" spellcheck="false">Line #84.    pp.run_time_dependent_model(model=model, params=solver_params)
</pre><p>Again, solver_params is redundant.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1218/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

File: tutorials/flux_discretizations.ipynb [hunk 19]
```diff
@@ -134,7 +134,7 @@
   },
```
Comment (IvarStefansson) [inline tutorials/flux_discretizations.ipynb:134]: <pre class="ql-syntax" spellcheck="false">Line #15.        "source": f, 
</pre><p>FYI, we have not enforced 88 character line length in tutorials.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1218/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

File: tutorials/mandels_problem.ipynb [hunk 20]
```diff
@@ -41,7 +41,7 @@
   },
```
Comment (IvarStefansson) [inline tutorials/mandels_problem.ipynb:41]: <pre class="ql-syntax" spellcheck="false">Line #2.    pp.run_time_dependent_model(model=setup)
</pre><p>Please fix output</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/1218/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

**Comment by IvarStefansson** [comment]
> We have been very pragmatic on this one, partly because we rarely use the history for notebooks, partly because reading the text version of notebooks is a pain. In my opinion, we should focus on the rendered part, and let the metadata be what it is. Do you agree @IvarStefansson?

Yes. But I also agree with Jakub, that 1) it's good practice not to add them if you haven't done any meaningful changes and 2) the tutorial which now renders an error message must be fixed.

**Comment by IvarStefansson** [comment]
Have you considered how this impact the line search classes?

File: tutorials/mandels_problem.ipynb (hunk 17 above)
Comment (mikeljordan) [reply tutorials/mandels_problem.ipynb]: The tutorial runs without error, can you please check again.

File: tests/functional/test_terzaghi.py [hunk 21]
```diff
@@ -156,9 +156,9 @@ def test_scaled_vs_unscaled_systems():
     }
     scaling = {"m": 0.001, "kg": 0.001}  # length in millimeters and mass in grams
     units = pp.Units(**scaling)
-    params_scaled = {"material_constants": material_constants_scaled, "units": units}
-    scaled = TerzaghiSetup(params=params_scaled)
-    pp.run_time_dependent_model(model=scaled, params=params_scaled)
+    model_params_scaled = {"material_constants": material_constants_scaled, "units": units}
+    scaled = TerzaghiSetup(params=model_params_scaled)
+    pp.run_time_dependent_model(model=scaled, params=model_params_scaled)
```
[the remark this answers is not in the document]
Comment (mikeljordan) [reply tests/functional/test_terzaghi.py:161]: If you look at the new review, yes it should.

File: tutorials/exporting_models.ipynb (hunk 8 above)
Comment (mikeljordan) [reply tutorials/exporting_models.ipynb]: This was used before the update on run_model() default argument for param.

File: tests/models/test_momentum_balance.py (hunk 7 above)
Comment (mikeljordan) [reply tests/models/test_momentum_balance.py]: I would suggest: model_params_ref

**Review by mikeljordan — COMMENTED** [review]
Some of the comments have been reviewed and resolved

File: tests/functional/test_terzaghi.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (jwboth) [reply tests/functional/test_terzaghi.py:161]: Can you explain? I oversaw this, and also do not understand why it is needed.

[the remark this answers is not in the document]
Comment (mikeljordan) [reply tests/functional/test_terzaghi.py]: No, it should not be passed. It was removed in the new updated. Previous reply was an oversight

File: tests/models/test_solution_strategy.py [hunk 22]
```diff
@@ -259,25 +259,26 @@ def bc_values_pressure(self, boundary_grid: pp.BoundaryGrid) -> np.ndarray:
...
-        "fracture_indices": [0, 1],
+    model_params = {
+        "fracture_indices": [0, 1],  
+        "full_rediscretization": True, 
         "cartesian": True,
         # Make flow problem non-linear:
-        "material_constants": {"fluid": pp.FluidConstants({"compressibility": 1})},
+        "material_constants": {"fluid": pp.FluidConstants({"compressibility": 1})}
     }
+    solver_params = {}  # {"fracture_indices": [0, 1]}
```
Comment (IvarStefansson) [inline tests/models/test_solution_strategy.py:269]:
```suggestion
    solver_params = {}
```

File: tutorials/equations.ipynb (hunk 9 above)
Comment (jwboth) [reply tutorials/equations.ipynb]: reverted now.

File: tutorials/diagnostics.ipynb (hunk 10 above)
Comment (jwboth) [reply tutorials/diagnostics.ipynb:575]: some changes added to file now, thus not reverted.\

**Review by jwboth — COMMENTED** [review]
From my side this good now. I noticed some places in the code which still used the old style and updated according to your suggested new style.

@IvarStefansson Would you mind checking the changes in the line search source code, and the solution strategy tutorial?

File: src/porepy/numerics/nonlinear/line_search.py [hunk 23]
```diff
@@ -18,7 +18,7 @@ class ConstraintLineSearchNonlinearSolver(
         LineSearchNewtonSolver,
     ):
         pass
-    model.params["nonlinear_solver"] = ConstraintLineSearchNonlinearSolver
+    solver_params["nonlinear_solver"] = ConstraintLineSearchNonlinearSolver
```
Comment (IvarStefansson) [inline src/porepy/numerics/nonlinear/line_search.py:21]: Please add an illustration of how the solver_params should be passed. Please also fix "model parameters"  to "solver parameters" in lines 12 and 24.

**Comment by IvarStefansson** [comment]
> @IvarStefansson Would you mind checking the changes in the line search source code, and the solution strategy tutorial?

I've had a look and made two minor comments. Otherwise all good!

**Comment by jwboth** [comment]
@mikeljordan, please tell me if you rather wish that I perform the changes, and I will do so.

**Comment by mikeljordan** [comment]
> @mikeljordan, please tell me if you rather wish that I perform the changes, and I will do so.

@jwboth yes I will do that.

**Comment by mikeljordan** [comment]
We have addressed all the comments and resolved the requested changes. The branch is now ready to be merged into develop.

**Review by jwboth — APPROVED** [review]
Thanks for the final changes, and the overall effort! @mikeljordan @isakhammer

**Pull request merged**